### PR TITLE
Small refactors of SBuiltin

### PR DESCRIPTION
**Pull request by carlpulley-da**
Refactor SBuiltin to:
- uniformly extend builtins with arity assertion checking
- uniformly extend builtins with token argument checking

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -197,42 +232,48 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
...
 
-  protected def execute(
+  /** On ledger builtins may reference the Speedy machine's ledger state.
+    *
+    * @param args arguments for executing the builtin
+    * @param machine the Speedy machine (machine state may be modified by the builtin)
+    * @param onLedger a reference to the Speedy machine's ledger state (which may be modified by the builtin)
+    * @return the builtin execution's resulting control value
+    */
+  protected def executeWithLedger(
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:266]: Renamed to better indicate the difference of usage here

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -37,6 +37,49 @@ import scala.jdk.CollectionConverters._
...
+  private[speedy] def execute(args: util.ArrayList[SValue], machine: Machine): Control
+
+  protected def crash(msg: String): Nothing =
+    throw SErrorCrash(getClass.getCanonicalName, msg)
+
+  protected def unexpectedType(i: Int, expected: String, found: SValue): Nothing =
+    crash(s"type mismatch of argument $i: expected $expected but got $found")
+}
+
+/** Extensions of this mixin trait ensure that builtin argument lists have a length matching the builtin's arity.
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:51]: Sorry, I don't like these mixin traits. I think they reduce, rather then improve, the code clarity. And I think they make understanding of the code execution harder to follow.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -37,6 +37,49 @@ import scala.jdk.CollectionConverters._
...
+  */
+private[speedy] trait BuiltinArityChecking { inner: SpeedyBuiltin =>
+
+  @inline
+  private[speedy] def execute(args: util.ArrayList[SValue], machine: Machine): Control = {
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:56]: These mixins don't work! -- The `execute` methods of both `BuiltinArityChecking` and `BuiltinTokenChecking` are never called.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:266]: :+1:

**Pull request closed without merging**